go/libraries/doltcore/sqle/database_provider.go

```diff
@@ -1007,20 +1007,103 @@ func isBranch(ctx context.Context, db SqlDatabase, branchName string, dialer dbf
 		return "", false, fmt.Errorf("unrecognized type of database %T", db)
 	}
 
+	brName, branchExists, err := isLocalBranch(ctx, ddbs, branchName)
+	if err != nil {
+		return "", false, err
+	}
+	if branchExists {
+		return brName, true, nil
+	}
+
+	brName, branchExists, err = isRemoteBranch(ctx, db, ddbs, branchName)
+	if err != nil {
+		return "", false, err
+	}
+	if branchExists {
+		return brName, true, nil
+	}
+
+	return "", false, nil
+}
+
+func isLocalBranch(ctx context.Context, ddbs []*doltdb.DoltDB, branchName string) (string, bool, error) {
 	for _, ddb := range ddbs {
-		branchName, branchExists, err := ddb.HasBranch(ctx, branchName)
+		brName, branchExists, err := ddb.HasBranch(ctx, branchName)
 		if err != nil {
 			return "", false, err
 		}
 
 		if branchExists {
-			return branchName, true, nil
+			return brName, true, nil
 		}
 	}
 
 	return "", false, nil
 }
 
+// isRemoteBranch is called only when the branch in connection string does not exist locally. It searches for a remote branch.
+// If there is a match, it creates a new local branch from the remote branch and sets its upstream to the remote branch.
+func isRemoteBranch(ctx context.Context, srcDB SqlDatabase, ddbs []*doltdb.DoltDB, branchName string) (string, bool, error) {
+	// parse branch name for possible remote name
+	remName := "origin"
+	brName := branchName
+
+	parts := strings.SplitN(branchName, dbRevisionDelimiter, 2)
+	if len(parts) > 1 {
+		remName = parts[0]
+		brName = parts[1]
+	}
+
+	for _, ddb := range ddbs {
+		remoteRefs, err := ddb.GetRemoteRefs(ctx)
+		if err == nil {
+			for _, rf := range remoteRefs {
+				if remRef, ok := rf.(ref.RemoteRef); ok && remRef.GetRemote() == remName && remRef.GetBranch() == brName {
+					err = createBranchOnDBAndSetUpstream(ctx, srcDB, ddb, brName, remRef.GetPath(), remRef)
+					if err != nil {
+						return "", false, err
+					}
+					return brName, true, nil
+				}
+			}
+		}
```

Comment:
(minor) It's probably worth adding a `HasRemoteTrackingBranch` function to`DoltDB` to mirror the existing `HasBranch` function. That would probably make the `isRemoteBranch` function look really similar to the `isLocalBranch` function, and it `HasRemoteTrackingBranch` seems like a useful method to have on `DoltDB`.